core/commonTypes.ml

```diff
@@ -0,0 +1,111 @@
+module Linearity = struct
+  type t = Any | Unl
+    [@@deriving eq,show]
+
+  let isAny lin = lin == Any
+  let isUnl lin = lin == Unl
+
+  let string_of = function
+    | Any -> "Any"
+    | Unl -> "Unl"
+end
+
+(* Convenient aliases for constructing values *)
+let linUnl = Linearity.Unl
+let linAny = Linearity.Any
+
+module DeclaredLinearity = struct
+  type t = Lin | Unl
+    [@@deriving show]
+
+  let isLin lin = lin == Lin
+  let isUnl lin = lin == Unl
+end
+
+(* Convenient aliases for constructing values *)
+let dlLin = DeclaredLinearity.Lin
+let dlUnl = DeclaredLinearity.Unl
+
+module Restriction = struct
+  type t =
+    | Any
+    | Base
+    | Session
+    | Effect
+    [@@deriving eq,show]
+
+  let isAny = function
+    | Any -> true
+    | _   -> false
+
+  let isBase = function
+    | Base -> true
+    | _    -> false
+
+  let isSession = function
+    | Session -> true
+    | _       -> false
+
+  let string_of = function
+    | Any     -> "Any"
+    | Base    -> "Base"
+    | Session -> "Session"
+    | Effect  -> "Eff"
+end
+
+(* Convenient aliases for constructing values *)
+let resAny     = Restriction.Any
```

Comment:
you're essentially introducing aliases for the more generically named variants like `Any`, by introducing `linAny`, `resAny`. In my opinion, this just shows that those constructors should be renamed. Effectively, only the aliases you define here are used in the rest of the code, so we could just rename the underlying constructors. What do the others think?

Reply:
> only the aliases you define here are used in the rest of the code

To be more precise: only aliases are used for constructing values. For deconstructing them (pattern matching) we still use constructors, for these types usually with explicit qualification. As for renaming the constructors to `ResAny`, `LinAny`, etc. - because I understand that's what you're proposing - I wonder what others think. Sam criticized this as poor man's namespaces, so I went with modules. Perhaps a middle ground here is shortening the names of modules to `Lin`, `DecLin`, `Res`, `PrKind`? We could then remove the aliases because spelling out the module name wouldn't be that painful.

Rant: I wish OCaml had pattern synonyms.

Reply:
I quite like the prefix, it is informative. But I wouldn't be opposed to the good ol' single letter prefix on constructor names either.